db/sovereign-db/src/ledger_db.rs

```diff
@@ -0,0 +1,245 @@
+        self.db.put::<SlotByNumber>(slot_number, slot)?;
+        self.db.put::<SlotByHash>(&slot.hash, slot_number)
+    }
+
+    fn put_batch(
+        &self,
+        batch: &StoredBatch,
+        batch_number: &BatchNumber,
+    ) -> Result<(), anyhow::Error> {
+        self.db.put::<BatchByNumber>(batch_number, batch)?;
+        self.db.put::<BatchByHash>(&batch.hash, batch_number)
+    }
+
+    fn put_transaction(
+        &self,
+        tx: &StoredTransaction,
+        tx_number: &TxNumber,
+    ) -> Result<(), anyhow::Error> {
+        self.db.put::<TxByNumber>(tx_number, tx)?;
+        self.db.put::<TxByHash>(&tx.hash, tx_number)
+    }
+
+    fn put_event(
+        &self,
+        event: &Event,
+        event_number: &EventNumber,
+        tx_number: TxNumber,
+    ) -> Result<(), anyhow::Error> {
+        self.db.put::<EventByNumber>(event_number, event)?;
+        self.db
+            .put::<EventByKey>(&(event.key.clone(), tx_number, event_number.clone()), &())
+    }
+
+    pub fn commit_slot(&self, data_to_commit: SlotCommit) -> Result<(), anyhow::Error> {
+        let item_numbers = {
+            let mut next_item_numbers = self.next_item_numbers.lock().unwrap();
+            let item_numbers = next_item_numbers.clone();
+
+            ensure!(
+                next_item_numbers.batch_number == data_to_commit.slot_data.batches.start.into(),
+                "First batch number must be the next in sequence."
+            );
+            if let Some(first_batch) = data_to_commit.batches.first() {
+                ensure!(
+                    next_item_numbers.tx_number == first_batch.txs.start.into(),
+                    "first ransaction number must be the next in sequence."
+                );
+            }
+            if let Some(first_tx) = data_to_commit.txs.first() {
+                ensure!(
+                    next_item_numbers.event_number == first_tx.events.start.into(),
+                    "first event number must be the next in sequence."
+                );
+            }
+
+            next_item_numbers.slot_number += 1;
+            next_item_numbers.batch_number += data_to_commit.batches.len() as u64;
+            next_item_numbers.tx_number += data_to_commit.txs.len() as u64;
+            next_item_numbers.event_number += data_to_commit.events.len() as u64;
+            item_numbers
+        };
+
+        // Insert data from "bottom up" to ensure consistency is present if the application crashes during insert
+
+        let mut event_number = item_numbers.event_number;
+        // Insert transactions and events
+        for (idx, (tx, event_group)) in data_to_commit
+            .txs
+            .into_iter()
+            .zip(data_to_commit.events.into_iter())
+            .enumerate()
+        {
+            let tx_number = TxNumber(item_numbers.tx_number + idx as u64);
+            for event in event_group.into_iter() {
+                self.put_event(&event, &EventNumber(event_number), tx_number)?;
```

Comment:
Nit: it looks like here we have full ownership of `EventNumber`, so we could pass it by value instead of by reference and avoid cloning inside `put_value`: https://github.com/Sovereign-Labs/sovereign/pull/151/files#diff-a2bb8d643ebc5191d070ef5ad4243cb92e2821b839b7b64f37f85bb1ad486b75R152 (unless the cloning is cheap but then we could derive `Copy`)

Reply:
It's already `Copy`!  I'll just change the `clone` into a dereference